### ISSUE TITLE
Help me debug compile performance with `eqx.Module` (with dropout)

### ISSUE THREAD

**Issue by carlesoctav** [patrick-kidger/equinox#1132]
For the past few weeks, I’ve been trying to figure out why my module (which uses dropout) takes about **200 seconds to compile**, while the Flax counterpart only takes **80 seconds**.
Both `train_step` functions generate almost the **same HLO** (in fact, the Equinox version is a bit simpler).

Previously, I ran into a similar issue when initializing a module inside a JIT function. Something like this, comparing Equinox and Flax:

```python
@jax.jit
def module():
    model = ModelClass(args, kwargs)
    # potentially apply sharding constraints
    # model = jax.lax.with_sharding_constraint(model, sharding_constraint)
    return model

@jax.jit
def flax_module():
    model = ModelClass()
    params = model.init(batch)
    # potentially apply sharding constraints
    # model = eqx.filter_shard(model, sharding_constraint)
    return params
```

When initializing a BERT model with dropout, it took about **80 seconds** for my Equinox version and **30 seconds** for Flax.
After dumping the HLO, I saw a lot of `split`, `reshape`, and other operations originating from `jax.random.split` in my module. I realized this was because I was doing **eager splitting** — splitting the key before passing it to every submodule, even for modules that didn’t use RNG directly (like “shell” modules that only contain submodules). I also wasn’t using `jax.random.fold_in()`.

After reading how Flax handles RNGs, I noticed that it does this **lazily** — basically only calling `jax.random.fold_in()` when necessary, which reduces RNG-related operations to roughly match the number of parameters.

After switching to this lazy RNG approach, compile time for model initialization dropped to about **20 seconds**, even faster than Flax. I think this is because Equinox is simpler at trace time (no scope object creation or shape computation).

However, this lazy RNG approach does **not** improve the **train_step compile time**.
The runtime is similar between Equinox and Flax, but compile time is still much slower for Equinox.

Compiletime
| Variant | With Dropout | Without Dropout |
| ------- | ------------ | --------------- |
| Equinox | 216 s        | 67 s            |
| Flax    | 78 s         | 38 s            |

Runtime (per step)
| Variant | With Dropout | Without Dropout |
| ------- | ------------ | --------------- |
| Equinox | 120 ms       | 100 ms          |
| Flax    | 150 ms       | 90 ms           |

After dumping the HLO again, they’re very similar. The number of lines and `_threefry` or `fold_in` ops are roughly the same.

What GPT-5-Codex said:

> Your latest HLO (`benchmark/new_train_hlo.txt`) is already on par with the Flax dropout variant:
>
> * Both graphs are around 1.2–1.3 MB (~12.7k vs 13.5k ops), and the dropout masks appear in the same places.
> * `_threefry*` ops dropped from 67 to 38, matching Flax.
> * The compiled body is still a single, partitioned train step (1 mesh axis, 4 partitions). XLA still has to fuse and partition thousands of ops, so ~200 s compile time is expected.
>
> The slowdown is no longer caused by RNG handling — it’s just the cost of compiling a full BERT + optimizer step.
>
> To reduce compile time further, try:
>
> * Staging the graph (e.g., JIT the forward/backward separately, or pre-JIT optimizer updates)
> * Using a smaller batch size or sequence length during compilation
> * Reusing an ahead-of-time (AOT) cache
>
> Once compiled, there shouldn’t be any per-step recompilation now that RNGs are properly folded in.

Earlier, the dropout version was significantly slower at runtime — around **220 ms vs 150 ms**. This was mostly due im using my custom `DArray` (an array with `pspec`, im still using it know, but now i'll unbox this before run the train_step) and fully realizing the dropout mask. After mimicking Flax’s behavior (using a broadcast axis for the mask) and later switching to lazy RNGs and `fold_in`, I reduced runtime to **120 ms**.

Even the non-dropout version started slower (**200 ms vs 90 ms**) but I managed to improve it to **130 ms**, and eventually **100 ms**.

Now, runtime is fine — my main concern is compile time with dropout.

You can find the HLO dumps and benchmark code here:
[https://github.com/carlesoctav/fme/benchmark/](https://github.com/carlesoctav/fme/benchmark/)

If you’d like me to run more experiments or variations (different mesh setups, staged compilation, etc.), I can do that.

Any thoughts or suggestions for improving compile time would be appreciated. Logs for each run are included below for reference.

logs (see the debugprint diff for the compile time):
```
eqx without dropout
(fme) carlesoctav@t1v-n-42336699-w-0:/mnt/carles/fme$ python benchmark/train.py
* Trackio project initialized: bert-mlm-fineweb
* Trackio metrics logged to: /mnt/carles/.cache/huggingface/trackio
* View dashboard by running in your terminal:
trackio show --project "bert-mlm-fineweb"
* or by running in Python: trackio.show(project="bert-mlm-fineweb")
* Created new run: dainty-sunset-0
make_module_opt took 0.0093s
DEBUGPRINT[44]: train.py:198: diff=68.05862800299656
Benchmarking:   0%|                                                                                                                                                                    | 0/101 [00:00<?, ?it/s][0,0,2159631] 2025-10-24 08:26:09,188 [benchmarking_utils.py] INFO Running step 0 (compilation step)...
[0,0,2159631] 2025-10-24 08:26:09,333 [benchmarking_utils.py] INFO Step 0 (compilation) took 0.1443s
Benchmarking: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 101/101 [00:21<00:00,  4.73it/s]
[0,0,2159631] 2025-10-24 08:26:20,708 [benchmarking_utils.py] INFO ==============================
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO ===== Benchmark Results ======
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO ==============================
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO Train Step Time (avg): 0.1044s
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO Train Step Time (median): 0.0926s
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO Train Step Time (std): 0.0644s
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO Next Batch Time (avg): 0.0060s
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO Batches/sec (avg): 9.58
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO Compile Time: 0.1443s
[0,0,2159631] 2025-10-24 08:26:20,709 [benchmarking_utils.py] INFO ==============================
[0,0,2159631] 2025-10-24 08:26:20,710 [benchmarking_utils.py] INFO Benchmark loop ended
(fme) carlesoctav@t1v-n-42336699-w-0:/mnt/carles/fme$ /usr/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 4 leaked shared_memory objects to clean up

=====================================
equinox dropout
make_module_opt took 0.0098s
DEBUGPRINT[44]: train.py:198: diff=216.35829847399145
Benchmarking:   0%|                                                                                                                                                                    | 0/101 [00:00<?, ?it/s][0,0,2171358] 2025-10-24 08:49:28,219 [benchmarking_utils.py] INFO Running step 0 (compilation step)...
[0,0,2171358] 2025-10-24 08:49:28,407 [benchmarking_utils.py] INFO Step 0 (compilation) took 0.1876s
Benchmarking: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 101/101 [00:23<00:00,  4.35it/s]
[0,0,2171358] 2025-10-24 08:49:41,587 [benchmarking_utils.py] INFO ==============================
[0,0,2171358] 2025-10-24 08:49:41,587 [benchmarking_utils.py] INFO ===== Benchmark Results ======
[0,0,2171358] 2025-10-24 08:49:41,587 [benchmarking_utils.py] INFO ==============================
[0,0,2171358] 2025-10-24 08:49:41,587 [benchmarking_utils.py] INFO Train Step Time (avg): 0.1265s
[0,0,2171358] 2025-10-24 08:49:41,587 [benchmarking_utils.py] INFO Train Step Time (median): 0.1155s
[0,0,2171358] 2025-10-24 08:49:41,587 [benchmarking_utils.py] INFO Train Step Time (std): 0.0617s
[0,0,2171358] 2025-10-24 08:49:41,588 [benchmarking_utils.py] INFO Next Batch Time (avg): 0.0018s
[0,0,2171358] 2025-10-24 08:49:41,588 [benchmarking_utils.py] INFO Batches/sec (avg): 7.91
[0,0,2171358] 2025-10-24 08:49:41,588 [benchmarking_utils.py] INFO Compile Time: 0.1876s
[0,0,2171358] 2025-10-24 08:49:41,588 [benchmarking_utils.py] INFO ==============================
[0,0,2171358] 2025-10-24 08:49:41,588 [benchmarking_utils.py] INFO Benchmark loop ended
(fme) carlesoctav@t1v-n-42336699-w-0:/mnt/carles/fme$ /usr/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 4 leaked shared_memory objects to clean


================================
Flax Dropout
* or by running in Python: trackio.show(project="benchmark")
* Created new run: bert-flax-with-dropout
DEBUGPRINT[45]: bert_flax_train.py:398: diff=78.56424835009966
[0,0,2166168] 2025-10-24 08:40:07,687 [bert_flax_train.py] INFO Running benchmark loop...
Benchmarking Flax:   0%|                                                                                                                                                               | 0/101 [00:00<?, ?it/s][0,0,2166168] 2025-10-24 08:40:17,735 [bert_flax_train.py] INFO Running step 0 (compilation step)...
[0,0,2166168] 2025-10-24 08:40:18,288 [bert_flax_train.py] INFO Step 0 (compilation) took 0.5532s
Benchmarking Flax: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 101/101 [00:26<00:00,  3.88it/s]
[0,0,2166168] 2025-10-24 08:40:33,720 [bert_flax_train.py] INFO ==================================================
[0,0,2166168] 2025-10-24 08:40:33,720 [bert_flax_train.py] INFO === Flax BERT Benchmark Results (WITH DROPOUT) ===
[0,0,2166168] 2025-10-24 08:40:33,720 [bert_flax_train.py] INFO ==================================================
[0,0,2166168] 2025-10-24 08:40:33,720 [bert_flax_train.py] INFO Train Step Time (avg): 151.3ms
[0,0,2166168] 2025-10-24 08:40:33,720 [bert_flax_train.py] INFO Train Step Time (median): 147.0ms
[0,0,2166168] 2025-10-24 08:40:33,721 [bert_flax_train.py] INFO Train Step Time (std): 41.7ms
[0,0,2166168] 2025-10-24 08:40:33,721 [bert_flax_train.py] INFO Next Batch Time (avg): 1.9ms
[0,0,2166168] 2025-10-24 08:40:33,721 [bert_flax_train.py] INFO Batches/sec (avg): 6.61
[0,0,2166168] 2025-10-24 08:40:33,721 [bert_flax_train.py] INFO Compile Time: 0.55s
[0,0,2166168] 2025-10-24 08:40:33,721 [bert_flax_train.py] INFO ==================================================
[0,0,2166168] 2025-10-24 08:40:33,721 [bert_flax_train.py] INFO Benchmark loop ended
[0,0,2166168] 2025-10-24 08:40:36,252 [bert_flax_train.py] INFO Benchmark completed!

================================
bert-flax-without dropout
DEBUGPRINT[45]: bert_flax_train.py:398: diff=35.554427047027275
[0,0,2200522] 2025-10-24 10:39:31,032 [bert_flax_train.py] INFO Running benchmark loop...
Benchmarking Flax:   0%|                                                                                                                                                                  | 0/101 [00:00<?, ?it/s][0,0,2200522] 2025-10-24 10:39:41,184 [bert_flax_train.py] INFO Running step 0 (compilation step)...
[0,0,2200522] 2025-10-24 10:39:41,584 [bert_flax_train.py] INFO Step 0 (compilation) took 0.3999s
Benchmarking Flax: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 101/101 [00:19<00:00,  5.08it/s]
[0,0,2200522] 2025-10-24 10:39:50,923 [bert_flax_train.py] INFO ==================================================
[0,0,2200522] 2025-10-24 10:39:50,923 [bert_flax_train.py] INFO === Flax BERT Benchmark Results (WITH DROPOUT) ===
[0,0,2200522] 2025-10-24 10:39:50,923 [bert_flax_train.py] INFO ==================================================
[0,0,2200522] 2025-10-24 10:39:50,923 [bert_flax_train.py] INFO Train Step Time (avg): 90.4ms
[0,0,2200522] 2025-10-24 10:39:50,924 [bert_flax_train.py] INFO Train Step Time (median): 85.8ms
[0,0,2200522] 2025-10-24 10:39:50,924 [bert_flax_train.py] INFO Train Step Time (std): 41.7ms
[0,0,2200522] 2025-10-24 10:39:50,924 [bert_flax_train.py] INFO Next Batch Time (avg): 2.1ms
[0,0,2200522] 2025-10-24 10:39:50,924 [bert_flax_train.py] INFO Batches/sec (avg): 11.07
[0,0,2200522] 2025-10-24 10:39:50,924 [bert_flax_train.py] INFO Compile Time: 0.40s
[0,0,2200522] 2025-10-24 10:39:50,924 [bert_flax_train.py] INFO ==================================================
[0,0,2200522] 2025-10-24 10:39:50,925 [bert_flax_train.py] INFO Benchmark loop ended
[0,0,2200522] 2025-10-24 10:39:53,572 [bert_flax_train.py] INFO Benchmark completed!


=======================================
```

anyway i was doing this on TPU V4

**Comment by patrick-kidger** [comment]
Hi there!
So my main advice would be to start bisecting through your program: starting from the program you already have, start simplifying it down (remove gradient steps, remove layers, ...) and see what is the minimal program for which your troublesome behaviour occurs.

Note that it's fairly unusual to JIT-compile a function with no arguments (in principle the output is a constant, and XLA may decide to constant-fold some amount of your program based on its own internal heuristics), and it's also fairly unusual to JIT-compile model initialisation (we often do that eagerly, as it usually just happens once at the start of training).

I hope that helps :)

**Comment by carlesoctav** [comment]
actually i solve it, i added a lifted version of scan (that can be applied to eqx.Module) and use scan instead of doing for loop.
